src/odbc/sql2tds.c

```diff
@@ -149,6 +149,129 @@ _odbc_blob_free(TDSCOLUMN *col)
         TDS_ZERO_FREE(col->column_data);
 }
 
+static TDS_INT
+odbc_convert_table(TDS_STMT * stmt, SQLTVP * src, TDS_TVP * dest, SQLLEN num_rows)
+{
+	SQLLEN i;
+	int j;
+	TDS_TVP_ROW * row;
+	TDS_TVP_ROW ** prow;
+	TDSPARAMINFO * params, * new_params;
+	TDS_DESC * apd, * ipd;
+	struct _drecord *drec;
+
+	dest->schema = src->schema;
+	dest->typename = src->typename;
+	dest->num_cols = src->num_cols;
+	dest->metadata = NULL;
+	dest->row = NULL;
+
+	/* Create a dummy row to store column metadata */
+	apd = desc_alloc(stmt, DESC_APD, SQL_DESC_ALLOC_AUTO);
+	ipd = desc_alloc(stmt, DESC_IPD, SQL_DESC_ALLOC_AUTO);
+
+	if ((dest->metadata = malloc(sizeof(TDS_TVP_ROW))) == NULL)
+		return -1;
+
+	dest->metadata->next = NULL;
+	params = NULL;
+	for (j = 0; j < src->num_cols; j++) {
+		new_params = tds_alloc_param_result(params);
+
+		if (j+1 > apd->header.sql_desc_count && desc_alloc_records(apd, j+1) != SQL_SUCCESS)
+			return -1;
+		drec = &apd->records[j];
+
+		if (odbc_set_concise_c_type(src->cols[j]->fCType, drec, 0) != SQL_SUCCESS)
+			return -1;
+
+		if (drec->sql_desc_type == SQL_C_CHAR || drec->sql_desc_type == SQL_C_WCHAR || drec->sql_desc_type == SQL_C_BINARY)
+			drec->sql_desc_octet_length = src->cols[j]->cbValueMax;
+		drec->sql_desc_indicator_ptr = (SQLLEN *) SQL_NULL_DATA;
+		drec->sql_desc_octet_length_ptr = (SQLLEN *) SQL_NULL_DATA;
+		drec->sql_desc_data_ptr = NULL;
+
+		/* field IPD related fields */
+		ipd = stmt->ipd;
+		if (j+1 > ipd->header.sql_desc_count && desc_alloc_records(ipd, j+1) != SQL_SUCCESS)
+			return -1;
+		drec = &ipd->records[j];
+
+		drec->sql_desc_parameter_type = src->cols[j]->fParamType;
+		if (odbc_set_concise_sql_type(src->cols[j]->fSqlType, drec, 0) != SQL_SUCCESS)
+			return -1;
+		if (src->cols[j]->fSqlType == SQL_DECIMAL || src->cols[j]->fSqlType == SQL_NUMERIC) {
+			drec->sql_desc_precision = src->cols[j]->cbColDef;
+			drec->sql_desc_scale = src->cols[j]->ibScale;
+		} else {
+			drec->sql_desc_length = src->cols[j]->cbColDef;
+		}
+
+		odbc_sql2tds(stmt, &ipd->records[j], &apd->records[j], new_params->columns[j], false, apd, 0);
+		params = new_params;
+	}
+	dest->metadata->params = params;
+
+	for (i = 0, prow = &dest->row; i < num_rows; prow = &(*prow)->next, i++) {
```

Comment:
Probably don't need `i` and just do something like
`for (prow = &dest->row; num_rows; prow = &(*prow)->next, num_rows--)`

Reply:
Turns out that `i` is being used to calculate the buffer offset for the i-th element, since all buffers are passed as byte pointers